Unify saved objects documentation into a single source of truth (#243788)
elastic/kibana: docs/extend/saved-objects.md

Do not use field mappings like you would use data types for the columns of a SQL database. Instead, field mappings are analogous to a SQL index. Only specify field mappings for the fields you wish to search on or query. By specifying `dynamic: false` in any level of your mappings, {{es}} will accept and store any other fields even if they are not specified in your mappings.

> Never use `enabled: false` or `index: false` in your mappings. {{es}} does not support toggling these mapping options, so if your plugin ever needs to query the data, you will not be able to do so. Since these fields cannot be queried, they would require migrating to a new field and making associated code changes. Instead, use `dynamic: false` which provides the same flexibility while maintaining the future ability to query fields if necessary.

Reviewer: ```suggestion
Never use `enabled: false` or `index: false` in your mappings. {{es}} does not support toggling these mapping options, so if your plugin ever needs to query the data, you will not be able to do so. Since these fields cannot be queried, they would require migrating to a new field and making associated code changes. Instead, use `dynamic: false` which provides the same flexibility while maintaining the future ability to query fields if necessary or do not specify a mapping for the field at all -- it will still be stored, just not mapped for search or aggregations.
```

        indexPattern: 'indexPattern1'
      },
      { references: [
          { id: '...', type: 'visualization', name: 'vis1' },
          { id: '...', type: 'index_pattern', name: 'indexPattern1' },
        ]
      }
    )
    ...
  }
);
```

[1] Note how `dashboard.panels[0].visualization` stores the name property of the reference (not the id directly) to be able to uniquely
identify this reference. This guarantees that the id the reference points to always remains up to date. If a
 visualization id was directly stored in `dashboard.panels[0].visualization` there is a risk that this id gets updated without
 updating the reference in the references array.

## Model versions

The modelVersion API is a new way to define transformations (*"migrations"*) for your savedObject types, and will
replace the "old" migration API after Kibana version `8.10.0` (where it will no longer be possible to register
migrations using the old system).

The main purpose of this API is to address two problems of the old migration system regarding managed ("serverless") deployments:

* savedObjects model versioning is coupled to the stack versioning (migrations are registered per stack version)

> * migration functions are not safe in regard to our BWC and ZDT requirements (Kibana N and N+1 running at the same time during upgrade)

Reviewer: ```suggestion
* migration functions are not safe in regard to our backwards-compatibility and and zero-downtime upgrade requirements (older versions of Kibana should continue to work as we run migraitons)
```

### How to remove a Saved Object type

#### Step 1: Remove the type registration

First, remove the type registration from your plugin. Also remove the type definition file and any references to it.

#### Step 2: Run the automated check

{{kib}} includes an automated check that detects when types are removed. Run it locally to update the `removed_types.json` file:

```bash
# Get your current commit ID
git log -n 1

# Get the merge-base commit with main
git merge-base -a <currentCommitSha> main

# Run the check with the baseline
node scripts/check_saved_objects --baseline <mergeBase> --fix
```

Replace `<currentCommitSha>` with your actual commit SHA from the first command, and `<mergeBase>` with the merge-base commit SHA from the second command.

This command will:

1. Compare the current registered types against the baseline (merge-base with main)
2. Detect any types that were removed
3. Automatically add the removed type name(s) to `removed_types.json`
4. Exit with an error message indicating which types were added

The error message will look like:

> ```log

Reviewer 2: ````suggestion
```shell
````

**Scenario 2:** I have only defined 1 new `modelVersion`. My PR was green before and all of a sudden it started failing :shrug:. The sanity checks on SO changes are performed against 2 baselines:

* Your PR merge base commit (or one of its ancestors). This should not change, so likely not the culprit.
* The current serverless release. If a high-severity issue occurs, Serverless might be rolled back to a previous version. This could in turn impact CI pipelines for PRs that would otherwise be fine. In other words, *"from the current serverless standpoint, your PR is not releasable at the moment"* . This should only happen exceptionally. Please take a look at the `#kibana-mission-control` channel to see if something is going on.
  * **Scenario 2.1**: {{kib}} has been rolled back. Please wait until the situation goes back to normal, i.e. an emergency release is performed after the rollback, and {{kib}} gets to a normal release state.
  * **Scenario 2.2**: {{kib}} has NOT been rolled back. Reach out to `#kibana-core` and we will help you figure out what's going on.

> ```log

Reviewer 2: ````suggestion
```shell
````
